### PR TITLE
Limit plugin

### PR DESCRIPTION
**Pull request by gazpachoking**
### Motivation for changes:
We needed a way to limit entries created by trakt_list plugin #2386, #2388. We added a limit option to that plugin. This is an experiment to see if we could use a more generic approach to limit any input plugin, while still allowing the optimization of not having the input plugin request more than is necessary from its source. (i.e. limiting HTTP or database requests)

### Detailed changes:
- Allows input plugins to be generators
- Adds 'limit' plugin, which takes a limit amount, as well as the configuration for another input plugin

### Config usage if relevant (new plugin or updated schema):
```
limit:
  amount: 5
  from:
    rss: http://blah.com
```

#### To Do:

- [x] Verify whether we like this idea
- [x] Convert input plugins which might take a long time to get many entries to generators
- [x] Deprecate limit options from plugins that implemented themselves (T411, Letterboxd)
- [x] Figure out how to best handle `@cached` decorator with generator inputs

### CODE REVIEW

File: flexget/plugins/input/limit.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+class PluginInputs(object):
+    """
+    Limits the number of entries an input plugin can produce.
+    """
+
+    schema = {
+        'type': 'object',
+        'properties': {
+            'amount': {'type': 'integer', 'minimum': -1},
+            'plugin': {
```
Comment (paranoidi) [inline flexget/plugins/input/limit.py]: Change to "from" to be in line with discover etc

Comment (gazpachoking) [reply flexget/plugins/input/limit.py]: Oh, good idea, that's much nicer.

**Comment by gazpachoking** [comment]
Looks like we like this, so I fixed up a few more things:
- Fixed the key for the subplugin to be called 'from'
- Switched over any input plugins I saw that were fetching successive pages to be generators
- Converted 'inputs' and 'sequence' plugins to both pass through generators

**Comment by gazpachoking** [comment]
@paranoidi @andrey-yantsen So, now the question is what to do with trakt_list plugin. Do we keep the limit option in there, or remove that again in favor of this? Do we keep the default limit for trakt_list, or maybe have a log message that recommends using this plugin if the number of entries is bigger than a certain amount?

**Comment by andrey-yantsen** [comment]
I suggeest removing the limit from trakt_list, and adding a check to an "upper" level, and if any input plugin except `limit` produces more than, let's say, 1000 items — send a notification to the log. Therefore we'll get the code with less cohesion level, where input plugins just produce items and don't care about amounts.

**Comment by andrey-yantsen** [comment]
BTW, feel free to steal my modified trakt_list from #2388, and convert it to the generator :)

**Comment by andrey-yantsen** [comment]
Shouldn't we deprecate `max_results` from `Letterboxd` and `T411` as well?

**Comment by gazpachoking** [comment]
@liiight limit_new is a bit different, that just limits the amount of accepted entries per run, and saves the rest for later runs

@andrey-yantsen You are right, those should probably be deprecated with this

**Comment by gazpachoking** [comment]
Hmm, things to figure out:
- What the hell changed about trakt_list_add test that it doesn't work anymore.
- How to handle our `@cached` input plugin decorator with generator style inputs. In memory persistence should be really easy, we can just use `itertools.tee` and store one copy of the resultant generator in cache. DB persistence on the other hand....

**Comment by andrey-yantsen** [comment]
> What the hell changed about trakt_list_add test that it doesn't work anymore.

Basically, you've fixed `TraktSet.clear()` method and now tests are making one request less :) previously it was populating items as the first step of the test, when the object is empty. Check out my latest commit to #2388.

**Comment by gazpachoking** [comment]
> > What the hell changed about trakt_list_add test that it doesn't work anymore.
> 
> Basically, you've fixed `TraktSet.clear()` method and now tests are making one request less :) previously it was populating items as the first step of the test, when the object is empty. Check out my latest commit to #2388.

Oh awesome, thanks for the investigation. :)

**Comment by gazpachoking** [comment]
@andrey-yantsen Is t411 a plugin anymore? I don't see it anywhere.

**Comment by andrey-yantsen** [comment]
Probably no - by some reason I’ve searched in wiki, not in github.

**Comment by cvium** [comment]
t411 as a site does not exist anymore

**Comment by andrey-yantsen** [comment]
from_digest plugin has a limit as well: https://github.com/Flexget/Flexget/blob/a4037da07f7d7707650596670c467ab3b7e748c9/flexget/plugins/operate/digest.py#L100

**Comment by gazpachoking** [comment]
Okay, I've modified the input caching to work with these new generator inputs. I think it's probably fine, but some feedback would be appreciated. Basic gist of how it works:
- Entries don't get cached to memory until they are iterated over. The memory cache also stores the generator though, so another task can iterate farther if the first task limited the input.
- Entries don't get cached to database at all, unless every entry is iterated over, since we can't cache the partially exhausted generator to db along with the partial entry list.

In practice, none of this probably matters, it's only protecting against the case of 2 tasks using the same input, but limited to a different quantity of entries (or increasing the limit number between runs while entries are cached to db.)

**Comment by gazpachoking** [comment]
Well, I haven't thought of any better ways to handle `@cached` decorator, so I've become a bit more confident on the solution I implemented. I think we are probably good to go on this PR unless anyone else has something.

**Comment by gazpachoking** [comment]
Going with it.

**Pull request merged**